src/qibolab/compilers/compiler.py

```diff
@@ -0,0 +1,171 @@
+from collections import defaultdict
+from dataclasses import dataclass, field
+
+import numpy as np
+from qibo import gates
+from qibo.config import raise_error
+from qibo.states import CircuitResult
+
+from qibolab.compilers.default import (
+    cz_rule,
+    identity_rule,
+    measurement_rule,
+    rz_rule,
+    u3_rule,
+    z_rule,
+)
+from qibolab.pulses import PulseSequence, ReadoutPulse
+
+
+class MeasurementMap(dict):
+    """Map from each measurement gate to the sequence of readout pulses implementing it."""
+
+    def __call__(self, backend, circuit, readout, nshots):
```

Comment:
Why do you use the `__call__` method, instead of giving it a name?

I can understand the `Compiler.__call__`, because an instance is a `compiler` that performs an action `compiler()` $\equiv$ `compiler.compile()`.
But here is a bit unexpected that a *map* object behaves like a function.

Reply:
> Why do you use the __call__ method, instead of giving it a name?

That is a valid question @AleCandido ;) After @stavros11 explanations, I also have my doubts about the naming here.  Besides the ability to store the mapping, the class has only one functionality, to register the result.  I thought that outside of the compiler it can be simply a function. Then the user can forget that it is actually a `dict` and use it for its sole purpose, to be called when needed.  Following this logic, calling the class maybe `RegisterMeasurementAction` or similar might be better.  Otherwise I am equally happy to make 'rename' `MeasurementMap.__call__`, e.g. `register(backend, )`.

Reply:
I agree that `__call__` is not the best name for this.

It is not entirely clear to me what is the final proposal for this. However, given that `MeasurementMap` is just a `dict` plus a function that takes a `backend`argument, but is only relevant for the `QibolabBackend`, another possibility would be to move `MeasurementMap.__call__` to `QibolabBackend.create_circuit_result` (or a different name). Then we completely drop the `MeasurementMap` object and in `Compiler.__call__` we create and return the `measurement_map` as a standard Python `dict` while in `QibolabBackend` we use this dict to create the `CircuitResult`.

The function would look like
```py
def create_circuit_result(self, measurement_map, circuit, readout, nshots):
    result = CircuitResult(backend, circuit, readout, nshots)
    for gate, sequence in measurement_map.items():
        _samples = map(lambda pulse: readout[pulse.serial].shots, sequence.pulses)
        samples = list(filter(lambda x: x is not None, _samples))
        gate.result.backend = self
        gate.result.register_samples(np.array(samples).T)
    return result
```
where now `self` is the `QibolabBackend` instance. @AleCandido @ingoroth let me know if you agree.

> I can understand the `Compiler.__call__`, because an instance is a `compiler` that performs an action `compiler()` ≡ `compiler.compile()`.

That's also a good point. I am fine with `compiler.compile()` if you prefer that.